File: frontend/src/services/skills.ts
// frontend/src/services/skills.ts
// 技能API服务 - 从后端获取技能数据

import type { Skill, SkillCategory, SkillListResponse, Era } from '@/types/skill'

const API_BASE = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api'

/**
 * 获取技能列表
 * @param era - 过滤年代 ('modern' | '1920s')
 * @param category - 过滤分类
 * @param search - 搜索关键词
 * @param includeSpecializations - 是否包含专攻技能
 */
export async function fetchSkills(
  era?: Era,
  category?: string,
  search?: string,
  includeSpecializations = false
): Promise<SkillListResponse> {
  const params = new URLSearchParams()

  if (era) params.append('era', era)
  if (category) params.append('category', category)
  if (search) params.append('search', search)
  if (includeSpecializations) params.append('include_specializations', 'true')

  const url = `${API_BASE}/skills${params.toString() ? `?${params.toString()}` : ''}`

  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(`Failed to fetch skills: ${response.statusText}`)
  }

  return response.json()
}

/**
 * 获取单个技能（通过ID）
 */
export async function fetchSkillById(id: number): Promise<Skill> {
  const response = await fetch(`${API_BASE}/skills/${id}`)
  if (!response.ok) {
    throw new Error(`Failed to fetch skill: ${response.statusText}`)
  }
  return response.json()
}

/**
 * 获取单个技能（通过名称，支持中英文）
 */
export async function fetchSkillByName(name: string): Promise<Skill> {
  const response = await fetch(`${API_BASE}/skills/name/${encodeURIComponent(name)}`)
  if (!response.ok) {
    throw new Error(`Failed to fetch skill: ${response.statusText}`)
  }
  return response.json()
}

/**
 * 获取技能分类列表
 */
export async function fetchSkillCategories(): Promise<SkillCategory[]> {
  const response = await fetch(`${API_BASE}/skills/categories`)
  if (!response.ok) {
    throw new Error(`Failed to fetch skill categories: ${response.statusText}`)
  }
  return response.json()
}

/**
 * 获取AI参考格式的技能信息（供LLM使用）
 */
export async function fetchSkillForAI(name: string): Promise<{
  id: number
  name: string
  name_en: string
  base_value: number
  category: string
  description: string | null
  difficulty_levels: string | null
  push_examples: string | null
  push_failure_examples: string | null
  opposing_skills: string | null
  specializations: string[]
}> {
  const response = await fetch(`${API_BASE}/skills/ai-reference/${encodeURIComponent(name)}`)
  if (!response.ok) {
    throw new Error(`Failed to fetch skill for AI: ${response.statusText}`)
  }
  return response.json()
}

/**
 * 工具函数：获取技能基础值（从API获取的技能列表中查找）
 */
export function getSkillBaseFromList(skills: Skill[], skillName: string): number {
  // 处理带特化的技能名（如 "格斗 (斗殴)" -> 找 "格斗"）
  const baseName = skillName.includes(' (') ? skillName.split(' (')[0] : skillName

  const skill = skills.find(s => s.name === baseName || s.name_en === baseName)
  return skill?.base_value ?? 0
}

/**
 * 工具函数：检查技能是否在指定年代可用
 */
export function isSkillAvailableInEra(skill: Skill, era: Era): boolean {
  return era === 'modern' ? skill.available_modern : skill.available_1920s
}

File: backend/src/api/skills.py
"""Skill API endpoints."""

from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import or_

from src.core.database import get_db
from src.models.skill import Skill, SkillCategory
from src.schemas.skill import (
    SkillCreate,
    SkillUpdate,
    SkillResponse,
    SkillListResponse,
    SkillCategoryResponse,
    SkillForAI,
)

router = APIRouter(prefix="/skills", tags=["skills"])


@router.get("", response_model=SkillListResponse)
def list_skills(
    era: Optional[str] = Query(None, description="Filter by era: 'modern' or '1920s'"),
    category: Optional[str] = Query(None, description="Filter by category"),
    search: Optional[str] = Query(None, description="Search by name"),
    include_specializations: bool = Query(False, description="Include specialization skills"),
    db: Session = Depends(get_db),
):
    """List all skills with optional filters."""
    import logging

    logger = logging.getLogger(__name__)
    logger.info(
        f"list_skills called: era={era}, category={category}, search={search}, include_specializations={include_specializations}"
    )

    try:
        query = db.query(Skill).options(joinedload(Skill.specializations))

        # Only show base skills (not specializations) by default
        if not include_specializations:
            query = query.filter(Skill.parent_skill_id == None)

        # Era filter
        if era == "modern":
            query = query.filter(Skill.available_modern == True)
        elif era == "1920s":
            query = query.filter(Skill.available_1920s == True)

        # Category filter
        if category:
            query = query.filter(Skill.category == category)

        # Search
        if search:
            search_term = f"%{search}%"
            query = query.filter(
                or_(Skill.name.ilike(search_term), Skill.name_en.ilike(search_term))
            )

        # Order by category, then name
        query = query.order_by(Skill.category, Skill.name)

        skills = query.all()
        total = len(skills)

        logger.info(f"Found {total} skills")

        return SkillListResponse(
            skills=[SkillResponse.model_validate(skill) for skill in skills], total=total
        )
    except Exception as e:
        logger.error(f"Error in list_skills: {e}", exc_info=True)
        raise


@router.get("/categories", response_model=List[SkillCategoryResponse])
def list_categories(db: Session = Depends(get_db)):
    """List all skill categories."""
    categories = db.query(SkillCategory).order_by(SkillCategory.sort_order).all()
    return [SkillCategoryResponse.model_validate(cat) for cat in categories]


@router.get("/name/{skill_name}", response_model=SkillResponse)
def get_skill_by_name(skill_name: str, db: Session = Depends(get_db)):
    """Get a skill by name (matches both Chinese and English names)."""
    skill = (
        db.query(Skill)
        .options(joinedload(Skill.specializations))
        .filter(or_(Skill.name == skill_name, Skill.name_en == skill_name))
        .first()
    )
    if not skill:
        raise HTTPException(status_code=404, detail="Skill not found")
    return SkillResponse.model_validate(skill)


@router.get("/ai-reference/{skill_name}", response_model=SkillForAI)
def get_skill_for_ai(skill_name: str, db: Session = Depends(get_db)):
    """Get detailed skill info for AI reference (matches both Chinese and English names)."""
    skill = (
        db.query(Skill).filter(or_(Skill.name == skill_name, Skill.name_en == skill_name)).first()
    )
    if not skill:
        raise HTTPException(status_code=404, detail="Skill not found")

    # Get specialization names
    spec_names = [s.name for s in skill.specializations]

    return SkillForAI(
        id=skill.id,
        name=skill.name,
        name_en=skill.name_en,
        base_value=skill.base_value,
        category=skill.category,
        description=skill.description,
        difficulty_levels=skill.difficulty_levels,
        push_examples=skill.push_examples,
        push_failure_examples=skill.push_failure_examples,
        opposing_skills=skill.opposing_skills,
        specializations=spec_names,
    )


@router.get("/{skill_id}", response_model=SkillResponse)
def get_skill(skill_id: int, db: Session = Depends(get_db)):
    """Get a single skill by ID."""
    skill = (
        db.query(Skill)
        .options(joinedload(Skill.specializations))
        .filter(Skill.id == skill_id)
        .first()
    )
    if not skill:
        raise HTTPException(status_code=404, detail="Skill not found")
    return SkillResponse.model_validate(skill)


@router.post("", response_model=SkillResponse)
def create_skill(skill_data: SkillCreate, db: Session = Depends(get_db)):
    """Create a new skill."""
    # Check if skill already exists
    existing = db.query(Skill).filter(Skill.name == skill_data.name).first()
    if existing:
        raise HTTPException(status_code=400, detail="Skill already exists")

    skill = Skill(**skill_data.model_dump())
    db.add(skill)
    db.commit()
    db.refresh(skill)
    return SkillResponse.model_validate(skill)


@router.put("/{skill_id}", response_model=SkillResponse)
def update_skill(skill_id: int, skill_data: SkillUpdate, db: Session = Depends(get_db)):
    """Update a skill."""
    skill = db.query(Skill).filter(Skill.id == skill_id).first()
    if not skill:
        raise HTTPException(status_code=404, detail="Skill not found")

    update_data = skill_data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(skill, key, value)

    db.commit()
    db.refresh(skill)
    return SkillResponse.model_validate(skill)


@router.delete("/{skill_id}")
def delete_skill(skill_id: int, db: Session = Depends(get_db)):
    """Delete a skill."""
    skill = db.query(Skill).filter(Skill.id == skill_id).first()
    if not skill:
        raise HTTPException(status_code=404, detail="Skill not found")

    db.delete(skill)
    db.commit()
    return {"message": "Skill deleted successfully"}
